hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestDirectoryDeletingServiceWithFSO.java

```diff
@@ -374,6 +376,128 @@ public void testDeleteFilesAndSubFiles() throws Exception {
+
+    /*    DirTable                               KeyTable
+    /v/b/snapDir                       /v/b/snapDir/testKey0 - testKey5
+    /v/b/snapDir/appRoot/              /v/b/snapDir/appRoot/parentDir0/childFile
+    /v/b/snapDir/appRoot/parentDir0/   /v/b/snapDir/appRoot/parentDir1/childFile
+    /v/b/snapDir/appRoot/parentDir1/   /v/b/snapDir/appRoot/parentDir2/childFile
+    /v/b/snapDir/appRoot/parentDir2/
+     */
+
+    Path root = new Path("/snapDir");
+    Path appRoot = new Path(root, "appRoot");
+    // Create  parent dir from root.
+    fs.mkdirs(root);
+
+    // Added 5 sub files inside root dir
+    for (int i = 0; i < 5; i++) {
+      Path path = new Path(root, "testKey" + i);
+      try (FSDataOutputStream stream = fs.create(path)) {
+        stream.write(1);
+      }
+    }
+
+    // Add 3 more sub files in different level
+    for (int i = 0; i < 3; i++) {
+      Path parent = new Path(appRoot, "parentDir" + i);
+      Path child = new Path(parent, "childFile");
+      ContractTestUtils.touch(fs, child);
+    }
+
+    KeyDeletingService keyDeletingService =
+        (KeyDeletingService) cluster.getOzoneManager().getKeyManager()
+            .getDeletingService();
+
+    // Before delete
+    assertTableRowCount(deletedDirTable, 0);
+    assertTableRowCount(keyTable, 8);
+    assertTableRowCount(dirTable, 5);
+
+    // Create snapshot
+    client.getObjectStore().createSnapshot(volumeName, bucketName, "snap1");
+    assertTableRowCount(snapshotInfoTable, 1);
+
+    // Case-1) Delete 3 Files directly.
+    for (int i = 0; i < 3; i++) {
+      Path path = new Path(root, "testKey" + i);
+      fs.delete(path, true);
+    }
+
+    DirectoryDeletingService dirDeletingService =
+        (DirectoryDeletingService) cluster.getOzoneManager().getKeyManager()
+            .getDirDeletingService();
+
+    // After delete. 5 more files left out under the root dir
+    assertTableRowCount(keyTable, 5);
+    assertTableRowCount(dirTable, 5);
+
+    // KeyDeletingService and DirectoryDeletingService will not
+    // clean up because the paths are part of a snapshot.
+    assertTableRowCount(deletedDirTable, 0);
+    assertTableRowCount(deletedKeyTable, 3);
+
+    assertSubPathsCount(dirDeletingService::getMovedFilesCount, 0);
+    assertSubPathsCount(dirDeletingService::getMovedDirsCount, 0);
+    assertSubPathsCount(dirDeletingService::getDeletedDirsCount, 0);
+
+    // Case-2) Delete dir
+    fs.delete(root, true);
+
+    // After delete. 5 sub files are still in keyTable.
+    // 4 dirs in dirTable.
+    assertTableRowCount(keyTable, 5);
+    assertTableRowCount(dirTable, 4);
+
+    // KeyDeletingService and DirectoryDeletingService will not
+    // clean up because the paths are part of a snapshot.
+    // As a result on 1 deleted dir and 3 deleted files will
+    // remain in dirTable and keyTable respectively.
+    assertTableRowCount(deletedDirTable, 1);
+    assertTableRowCount(deletedKeyTable, 3);
```

Comment:
True. Though I think it's better that we trigger both KDT and DDT here and wait for their completion, just to verify that even if the tasks are run none of those are cleaned up.

Reply:
I manually tested for `assertTableRowCount(deletedKeyTable, 2);` to see if keys are deleted, but it never was deleted. Did the same for `deletedDirTable`